active-rfcs/0000-relative-url-scheme.md

```diff
@@ -0,0 +1,95 @@
+- Start Date: 2021-12-20
+- Reference Issues: [#46](https://github.com/withastro/rfcs/pull/46)
+- Implementation PR: <!-- leave this empty -->
+
+# Summary
+
+Astro should support source-relative URL resolution in the HTML block of Astro files.
+
+# Example
+
+To resolve the source-relative URL of an image, use a `local:` prefixed URL.
+
+```html
+<img src="local:kitten.avif" alt="Kitten" />
+```
+
+# Motivation
+
+In `.astro` files, it’s not always easy to know what a relative URL is relative to.
+
+```astro
+<img src="kitten.avif" alt="kitty" />
+```
+
+In this example, is `kitten.avif` relative to the output `.html` or the source `.astro`? Or, if this example is found within a component `.astro` file, is it relative to that component file?
+
+This ambiguity raises a need for:
+
+- An intuitive way to author URLs relative to a source file.
+- An intuitive way to author URLs within the HTML of an `.astro` file.
+
+# Detailed design
+
+A `local` scheme instructs Astro to resolve the URL relative to the source file.
+
+```astro
+<img srcset="local:kitten.avif 1x, local:kitten@2x.avif" alt="Kitten" />
+```
+
+```astro
+<picture>
+  <source srcset="local:media/kitten@2x.avif" media="(min-width: 800px)" />
+  <img src="local:media/kitten.jpg" alt="" />
+</picture>
+```
+
+```astro
+<div style="background-image: url(local:kitten-background.avif);"></div>
```

Comment:
I like how this idea, rather than `local:` as an attribute prefix, allows you to use it anywhere and Astro doesn't need to special-case certain usage.

Reply:
I appreciate this as well, but it also exposes a concern. Would this be supported in other files within an `.astro` project? If so, how could we do that without scanning the final HTML output?

Reply:
I would expect any code that is currently run thru our resolver to get picked up by this, with the addition of attributes by this RFC.

Perhaps I can flesh out a `resolveId` prototype for Vite, to demonstrate how the functionality could be paired in other compiled files.

Related: Does `resolveId` handle `<script>import X from '@some/npm-module'</script>`? These imports are currently resolved in client-side code, but I wasn’t sure by what mechanism.

Reply:
I think this needs to be scoped to just .astro files. It should happen in the compiler by replacing `'local:./path/to/something.js'` with `{await import('./path/to/something.js?url')}`

Reply:
> I think this needs to be scoped to just `.astro` files.

Thank you for the design detail. I was expecting this to work within any file within `src`.

Related question: Would you expect `local:` to work within `<script></script>` or `<style><style>` within `.astro`?

```astro
<style>
:host {
  background: url(local:kitten.jpg);
}
</style>
```

Reply:
Is it possible to keep the original scope of only introducing a `local:src` directive? That would seem to meet the goals of the motivation section, so it's not clear to me why we would need to increase the scope to support `url(...)` as well.

Also I don't believe that this is required: Vite says that it resolves `url()` already (src-relative), similar to how CSS `@import` and ESM `import` work.

Reply:
I can see how that would be useful, but a couple of problems:

- This overlaps with usecases for `define:vars`
- Static styles and hoisted scripts have to be static when .astro files are compiled, this would break that.

I would probably recommend scoping this proposal to attribute values to avoid hitting these types of issues.

This is a good reason to prefer `local:src` attribute prefix instead, as well. Although that has the downsides already discussed.

Reply:
Hopefully these are resolved by aa6df9d5fa30d5086f5562b205a2015f69939721